font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,168 @@
 // limitations under the License.
 //
 //! Rasterization powered by DirectWrite
-use dwrote::{FontCollection, FontStretch, FontStyle, FontWeight, GlyphOffset, GlyphRunAnalysis};
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::fmt::{self, Display, Formatter};
+use std::os::windows::ffi::OsStringExt;
+
+use dwrote::{
+    FontCollection, FontFace, FontFallback, FontStretch, FontStyle, FontWeight, GlyphOffset,
+    GlyphRunAnalysis, TextAnalysisSource, TextAnalysisSourceMethods, DWRITE_GLYPH_RUN,
+};
+
+use winapi::shared::ntdef::{HRESULT, LOCALE_NAME_MAX_LENGTH};
+use winapi::um::dwrite;
+use winapi::um::winnls::GetUserDefaultLocaleName;
 
 use super::{
     BitmapBuffer, FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight,
 };
 
+/// This struct contains the properties from dwrote::Font needed in the rest of this module.
+///
+/// We use this struct rather than dwrote::Font for performance reasons: the getter methods on
+/// dwrote::Font need to do things like String conversion from C++ widestrings to Rust's
+/// UTF-8 strings.
+pub struct Font {
+    face: FontFace,
+    family_name: String,
+    weight: FontWeight,
+    style: FontStyle,
+    stretch: FontStretch,
+}
+
 pub struct DirectWriteRasterizer {
-    fonts: Vec<dwrote::FontFace>,
+    fonts: HashMap<FontKey, Font>,
+    keys: HashMap<FontDesc, FontKey>,
     device_pixel_ratio: f32,
+    collection: FontCollection,
+    fallback: Option<FontFallback>,
```

Comment:
The naming for this is a bit awkward too. Maybe there's a name that would better describe it than just 'fallback'?

Reply:
Similar to above, the name `fallback` leads directly from the directwrite `FontFallback` concept.

`fallback_sequences` ? (Inspired by the description of [FontFallback at Microsoft](https://docs.microsoft.com/en-gb/windows/win32/api/dwrite_2/nn-dwrite_2-idwritefontfallback))

Reply:
Seems like it would be `fallback_sequence`, without the `s`?

The word `sequence` seems almost a bit hard to grasp for me, but I think it's a decent choice that should work just fine.